aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -1872,6 +1894,149 @@ Tensor orgqr(const Tensor& input, const Tensor& tau) {
   return at::linalg_householder_product(input, tau);
 }
 
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+DEFINE_DISPATCH(ormqr_stub);
+
+void ormqr_out_helper(const Tensor& input, const Tensor& tau, const Tensor& other, const Tensor& result, bool left, bool transpose) {
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.dim() >= 2);
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(other.dim() >= 2);
+
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(other.size(left ? -2 : -1) >= tau.size(-1));
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(other.size(left ? -2 : -1) == input.size(-2));
+
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.scalar_type() == tau.scalar_type());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.device() == tau.device());
+
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.scalar_type() == other.scalar_type());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.device() == other.device());
+
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(result.scalar_type() == input.scalar_type());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(result.device() == input.device());
+
+  // if 'result' has no elements we can modify it
+  if (result.numel() == 0) {
+    at::native::resize_as_(result, other.transpose(-2, -1), MemoryFormat::Contiguous);
+    result.transpose_(-2, -1);
+  }
+
+  // 'result' tensor must be in batched column major order (Fortran contiguous)
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(result.transpose(-2, -1).is_contiguous());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(result.sizes().equals(other.sizes()));
+
+  // 'tau' tensor must be contiguous
+  Tensor tau_ = tau;
+  if (!tau.is_contiguous()) {
+    tau_ = at::empty(tau.sizes(), tau.options(), MemoryFormat::Contiguous);
+    tau_.copy_(tau);
+  }
+
+  // 'input' tensor must be Fortran contiguous
+  Tensor input_ = input;
+  if (!input.transpose(-2, -1).is_contiguous()) {
+    input_ = at::empty(input.transpose(-2, -1).sizes(), input.options(), MemoryFormat::Contiguous);
+    input_.transpose_(-2, -1);
+    input_.copy_(input);
+  }
+
+  // ormqr_stub (apply_ormqr) performs calculations in-place and 'result' must be a copy of 'other'
+  result.copy_(other);
+
+  ormqr_stub(result.device().type(), input_, tau_, result, left, transpose);
+}
+
+Tensor& ormqr_out(const Tensor& input, const Tensor& tau, const Tensor& other, bool left, bool transpose, Tensor& result) {
+  TORCH_CHECK(input.dim() >= 2, "torch.ormqr: input must have at least 2 dimensions.");
+  TORCH_CHECK(other.dim() >= 2, "torch.ormqr: other must have at least 2 dimensions.");
+
+  int64_t left_size_condition = left ? -2 : -1;
+  TORCH_CHECK(
```

Comment:
missing `input.size(-2) >= input.size(-1)`?

Reply:
No, as you see from the tests passing this function works both for `m >= n` and `m < n` matrices.
For the case with `m < n` only the first _m_ columns that represent _m_ Householder vectors are used by LAPACK.

```python
In [1]: import torch
In [2]: a = torch.randn(3, 5)
In [3]: h, tau = torch.geqrf(a)
In [4]: h.shape
Out[4]: torch.Size([3, 5])
In [5]: c = torch.randn(3, 7)
In [6]: res = torch.ormqr(h, tau, c)
In [7]: q, _ = torch.linalg.qr(a)
In [8]: torch.allclose(q @ c, res)
Out[8]: True
```

Reply:
Then we should add that the case `m < n` to the docs.

Reply:
Why do you think the case `m < n` is special?

Reply:
Because in `householder_product` is not even considered:
https://pytorch.org/docs/master/generated/torch.linalg.householder_product.html

Reply:
Well, `householder_product` is related but it's a different function. It has the same constraints on the input as the original orgqr implementation had:
https://github.com/pytorch/pytorch/blob/27f7d1c2865355c694fb964609df45974748615b/aten/src/TH/generic/THTensorLapack.cpp#L341
and LAPACK's orgqr has this `m>=n` requirement, but n there is the number of columns of Q to be computed, not `n=input.shape[-1]`. It has a side effect that the output of `torch.geqrf` can't be directly used with `torch.orgqr`, but it can be used with `ormqr` with the current implementation.

```python
In [1]: import torch
In [2]: a = torch.randn(3, 5)
In [3]: h, tau = torch.geqrf(a)
In [4]: c = torch.eye(3)
In [5]: torch.ormqr(h, tau, c) # narrow of `h` is not required
Out[5]: 
tensor([[-0.1468,  0.8266, -0.5433],
        [-0.0029, -0.5496, -0.8354],
        [-0.9892, -0.1211,  0.0831]])
In [6]: torch.linalg.qr(a)[0]
Out[6]: 
tensor([[-0.1468,  0.8266, -0.5433],
        [-0.0029, -0.5496, -0.8354],
        [-0.9892, -0.1211,  0.0831]])
In [7]: torch.linalg.householder_product(h, tau)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-7-70a58677659d> in <module>
----> 1 torch.linalg.householder_product(h, tau)

RuntimeError: torch.linalg.householder_product: input.shape[-2] must be greater than or equal to input.shape[-1]
In [8]: torch.linalg.householder_product(h.narrow(-1, 0, 3), tau) # narrow is required here
Out[8]: 
tensor([[-0.1468,  0.8266, -0.5433],
        [-0.0029, -0.5496, -0.8354],
        [-0.9892, -0.1211,  0.0831]])
```

Reply:
I do not have any strong opinions on how to handle this, as I really think that these functions are too low level to form part of the python API.

That being said, it's quite annoying that we have an `h, tau` decomposition in several functions, and that each of them has slightly different requirements.

Reply:
Even more, the `seealso` section of `householder_reflection` should be updated to reflect this dissonance.

Reply:
In every place where we have a multiplication of Q with some other matrix, using this function should be more efficient. In PyTorch Python code, there is for example one place it could be used, in the implementation of `lobpcg` the Q matrix is explicitly formed and then used in multiplication:
https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/torch/_lobpcg.py#L878
https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/torch/_linalg_utils.py#L80-L88

To make using this function more accessible, the documentation should be improved, the name should be changed and autograd support added. If we do that, we can revisit the constraints on the input and make it consistent.

Reply:
That's spot on. Let's do that in a follow-up after the branch-cut.